src/lazy/binary/immutable_buffer.rs

```diff
@@ -0,0 +1,939 @@
+                .map_err(|_| decoding_error_raw("found a field id with length over 255 bytes"))?;
+            (
+                Some(field_id_var_uint.value()),
+                field_id_length,
+                input_after_field_id,
+            )
+        } else {
+            (None, 0, initial_input)
+        };
+
+        let mut annotations_header_length = 0u8;
+        let mut annotations_sequence_length = 0u8;
+        let mut expected_value_length = None;
+
+        let mut type_descriptor = input.peek_type_descriptor()?;
+        if type_descriptor.is_annotation_wrapper() {
+            let (wrapper, input_after_annotations) =
+                input.read_annotations_wrapper(type_descriptor)?;
+            annotations_header_length = wrapper.header_length;
+            annotations_sequence_length = wrapper.sequence_length;
+            expected_value_length = Some(wrapper.expected_value_length);
+            input = input_after_annotations;
+            type_descriptor = input.peek_type_descriptor()?;
+            if type_descriptor.is_annotation_wrapper() {
+                return decoding_error("found an annotations wrapper ");
+            }
+        } else if type_descriptor.is_nop() {
+            (_, input) = input.consume_nop_padding(type_descriptor)?;
+        }
+
+        let header = type_descriptor
+            .to_header()
+            .ok_or_else(|| decoding_error_raw("found a non-value in value position"))?;
+
+        let header_offset = input.offset();
+        let (length, _) = input.consume(1).read_value_length(header)?;
+        let length_length = u8::try_from(length.size_in_bytes()).map_err(|_e| {
+            decoding_error_raw("found a value with a header length field over 255 bytes long")
+        })?;
+        let value_length = length.value(); // ha
+        let total_length = field_id_length as usize
+            + annotations_header_length as usize
+            + 1 // Header byte
+            + length_length as usize
+            + value_length;
+
+        if let Some(expected_value_length) = expected_value_length {
+            let actual_value_length = 1 + length_length as usize + value_length;
+            if expected_value_length != actual_value_length {
+                println!("{} != {}", expected_value_length, actual_value_length);
+                return decoding_error(
+                    "value length did not match length declared by annotations wrapper",
+                );
+            }
+        }
+
+        let encoded_value = EncodedValue {
+            header,
+            field_id_length,
+            field_id,
+            annotations_header_length,
+            annotations_sequence_length,
+            header_offset,
+            length_length,
+            value_length,
+            total_length,
+        };
+        let lazy_value = LazyRawValue {
+            encoded_value,
+            input: initial_input,
+        };
+        Ok(Some(lazy_value))
```

Comment:
Only the header (field ID, annotations, opcode, length) must be in the buffer for this to return a `LazyRawValue`; those fields are enough to construct an `EncodedValue` that retains their respective offsets. If the body of the value is incomplete (only half is in the buffer, eg), then calling `read()` will return an `Incomplete` error. A blocking wrapper for this would need to handle the `Incomplete` by providing more data and trying again.